[8.x] Create MigrateAutoCommand.php

redbastie opened the pull request:
# How It Works

The `migrate:auto` command will automatically diff the database and apply the necessary changes via Doctrine each time it is executed.

## Usage

Specify a `migration` method in your Laravel models:

    class Vehicle extends Model
    {
        use HasFactory;
    
        public function migration(Blueprint $table)
        {
            $table->id();
            $table->string('name');
            $table->timestamps();
        }
    }

Run the `migrate:auto` command:

    php artisan migrate:auto --fresh --seed --force

Note that `--fresh`, `--seed`, and `--force` are optional.

## Traditional Migrations

The `migrate:auto` command will run the traditional migration files before the automatic migration methods.

## Model Namespace

This would require a change to `laravel/laravel`, adding a `model_namespace` value to the `database` config file:

    /*
    |--------------------------------------------------------------------------
    | Model Namespace
    |--------------------------------------------------------------------------
    |
    | Here you may specify the namespace where your model classes reside. 
    | This is what will be used when resolving the automatic migration methods 
    | within your application when running the migrate:auto command.
    |
    */

    'model_namespace' => 'App\\Models',

redbastie committed:
Create MigrateAutoCommand.php

# How It Works

The `migrate:auto` command will automatically diff the database and apply the necessary changes via Doctrine each time it is executed.

## Usage

Specify a `migration` method in your Laravel models:

    class Vehicle extends Model
    {
        use HasFactory;
    
        public function migration(Blueprint $table)
        {
            $table->id();
            $table->string('name');
            $table->timestamps();
        }
    }

Run the `migrate:auto` command:

    php artisan migrate:auto --fresh --seed --force

Note that `--fresh`, `--seed`, and `--force` are optional.

## Traditional Migrations

The `migrate:auto` command will run the traditional migration files before the automatic migration methods.

redbastie committed:
Update MigrateAutoCommand.php

GrahamCampbell commented on src/Illuminate/Foundation/Console/MigrateAutoCommand.php:
```diff
@@ -0,0 +1,63 @@
+        Artisan::call('migrate' . ($this->option('fresh') ? ':fresh' : null) . ($this->option('force') ? ' --force' : null));
+
+        $filesystem = new Filesystem;
+        $dir = base_path('app/Models');
```
What if the models are no there.

redbastie replied:
This would only work for the default model path that comes with L8. However, we could add a config value to the `database` config file, something like `model_path`?

GrahamCampbell commented on src/Illuminate/Foundation/Console/MigrateAutoCommand.php:
```diff
@@ -0,0 +1,63 @@
+
+        if ($filesystem->exists($dir)) {
+            foreach ($filesystem->allFiles($dir) as $file) {
+                $class = app('App\\Models\\' . str_replace(['/', '.php'], ['\\', ''], $file->getRelativePathname()));
```
What if this is not the namespace

redbastie replied:
See my previous comment. If we added a `model_path` to the `database` config file, we could use that to determine the model path and it's appropriate namespace. Thoughts?

kafkiansky commented on src/Illuminate/Foundation/Console/MigrateAutoCommand.php:
```diff
@@ -0,0 +1,63 @@
+            foreach ($filesystem->allFiles($dir) as $file) {
+                $class = app('App\\Models\\' . str_replace(['/', '.php'], ['\\', ''], $file->getRelativePathname()));
+
+                if (method_exists($class, 'migration')) {
```
OMG, model knows about factories, now they will know about migrations? What about versioning, if model would be change? What about migration's drop?

redbastie replied:
Doctrine handles all of that via the `diffTable` method.  https://github.com/doctrine/dbal/blob/2.12.x/lib/Doctrine/DBAL/Schema/Comparator.php

kafkiansky replied:
How you plan to drop the initial migration? What if on project we use a module architecture? `model_path` option in config  would be useless. You must add an option in command that ask to choose models path, if you want, or use default.

redbastie replied:
You can still use this alongside traditional migrations for more complex applications. It runs the traditional migrations before the automatic migration methods. We could also add a `--path` option to the command as well, in case you want to specify the model path for the command to use in edge cases.

redbastie committed:
Update composer.json

redbastie committed:
use db model_path config

redbastie commented:
I updated the command to use the `model_path` value inside of the `database` config file. We would need to add this to `config/database.php` of the `laravel/laravel` repo as well obviously:

    /*
    |--------------------------------------------------------------------------
    | Model Path
    |--------------------------------------------------------------------------
    |
    | Here you may specify where your model classes reside. This is what
    | will be utilized when resolving the automatic migration methods within
    | your application when running the migrate:auto command.
    |
    */

    'model_path' => 'app/Models',

redbastie committed:
Update composer.json

redbastie committed:
make styleci happy

redbastie committed:
model_path missing abort

Requires `model_path` to be present in the `config/database.php` file:

    /*
    |--------------------------------------------------------------------------
    | Model Path
    |--------------------------------------------------------------------------
    |
    | Here you may specify where your model classes reside. This is what
    | will be utilized when resolving the automatic migration methods within
    | your application when running the migrate:auto command.
    |
    */

    'model_path' => 'app/Models',

redbastie committed:
change config model_path to model_namespace

Adds consistency with other config values:

    /*
    |--------------------------------------------------------------------------
    | Model Namespace
    |--------------------------------------------------------------------------
    |
    | Here you may specify the namespace where your model classes reside. 
    | This is what will be used when resolving the automatic migration methods 
    | within your application when running the migrate:auto command.
    |
    */

    'model_namespace' => 'App\\Models',

redbastie committed:
styleci happiness

redbastie committed:
styleci happiness

redbastie committed:
styleci happiness

taylorotwell commented:
This could make for an interesting package.
